### PR TITLE
Issue 11 entry serialization type

### PR DESCRIPTION
**Pull request by LikeAnyan**
Added `@EntryClass` annotation that is used in `@ProviderKey` annotation to tell type that should be used for object serialization & deserialization. Pr links this issue: https://github.com/epam/CoroutinesCache/issues/11.
`@EntryClass` annotation has two params - first called rawType, kotlin class should be passed there, the second one - called type params with type vararg of EntryClass (used as type arguments for ParameterizedType). Here's usage examples:

```
@ProviderKey("key1", EntryClass(Data::class))
fun getData(data: Deferred<Data>): Deferred<Data>

@ProviderKey("key2", EntryClass(List::class, EntryClass(Data::class)))
fun getListData(data: Deferred<List<Data>>): Deferred<List<Data>>`

@ProviderKey("key3", EntryClass(Map::class, EntryClass(Int::class), EntryClass(Data::class)))
fun getMapData(data: Deferred<Map<Int, Data>>): Deferred<Map<Int, Data>>
```

Please not that this PR also includes: https://github.com/epam/CoroutinesCache/pull/13

### CODE REVIEW

File: core/src/main/java/com/epam/coroutinecache/core/DiskCache.kt [hunk 1]
```diff
@@ -66,44 +66,17 @@ class DiskCache(
...
-                        typeRecord = jsonMapper.newParameterizedType(Record::class.java, typeMap)
-                    }
-                    else -> {
-                        typeRecord = jsonMapper.newParameterizedType(Record::class.java, dataClassName)
-                    }
-                }
-                val diskRecord: Record<T>? = jsonMapper.fromJson(resultedFile.absoluteFile, typeRecord)
+                val type = jsonMapper.newParameterizedType(Record::class.java, entryType)
+                val diskRecord: Record<T>? = jsonMapper.fromJson(resultedFile, type)
                 diskRecord?.sizeOnMb = (resultedFile.length().toFloat() / 1024 / 1024)
```
Comment (vvsevolodovich) [inline core/src/main/java/com/epam/coroutinecache/core/DiskCache.kt]: let's extract the constant here

File: core/src/main/java/com/epam/coroutinecache/core/DiskCache.kt [hunk 2]
```diff
@@ -66,44 +66,17 @@ class DiskCache(
...
-                return diskRecord
+                diskRecord
             } catch (exception: Exception) {
-                return null
```
Comment (vvsevolodovich) [inline core/src/main/java/com/epam/coroutinecache/core/DiskCache.kt:106]: We need to log those exceptions, let's create a separate issue for Logging

File: core/src/main/java/com/epam/coroutinecache/core/Memory.kt [hunk 3]
```diff
@@ -8,15 +8,15 @@ interface Memory {
      * @param key The key associated with the Record to be retrieved from memory
      * @see Record
      */
-    fun <T> getRecord(key: String): Record<T>?
+    fun <T: Any> getRecord(key: String): Record<T>?
```
Comment (vvsevolodovich) [inline core/src/main/java/com/epam/coroutinecache/core/Memory.kt]: what is the point of extending Any?

**Review by vvsevolodovich — APPROVED** [review]

**Review by IvanSakharovskii — APPROVED** [review]

File: core/src/main/java/com/epam/coroutinecache/core/Memory.kt (hunk 3 above)
Comment (LikeAnyan) [reply core/src/main/java/com/epam/coroutinecache/core/Memory.kt]: > what is the point of extending Any?

My bad, extending Any came from my playground

File: core/src/main/java/com/epam/coroutinecache/core/DiskCache.kt (hunk 1 above)
Comment (LikeAnyan) [reply core/src/main/java/com/epam/coroutinecache/core/DiskCache.kt]: > let's extract the constant here

what do you mean by the constant?

File: core/src/main/java/com/epam/coroutinecache/core/DiskCache.kt [hunk 4]
```diff
@@ -66,44 +66,17 @@ class DiskCache(
...
-                        val classKeyMap = Class.forName(tempData?.getDataKeyMapClassName())
-                        val typeMap = jsonMapper.newParameterizedType(collectionClassName, classKeyMap, dataClassName)
-                        typeRecord = jsonMapper.newParameterizedType(Record::class.java, typeMap)
-                    }
-                    else -> {
-                        typeRecord = jsonMapper.newParameterizedType(Record::class.java, dataClassName)
-                    }
-                }
-                val diskRecord: Record<T>? = jsonMapper.fromJson(resultedFile.absoluteFile, typeRecord)
+                val type = jsonMapper.newParameterizedType(Record::class.java, entryType)
```
Comment (LikeAnyan) [inline core/src/main/java/com/epam/coroutinecache/core/DiskCache.kt:74]: @vlivanov, @IvanSakharovskii 
Currently here the only one place where newParameterizedType from jsonMapper interface used (just forgot to replace).
obtainTypeFromAnnotation method relies on methods from our Types object and I don't know any reason to keep this method in JsonMapper interface, so I suggest to remove this method from JsonMapper interface and replace with Type's object method call. Or rely only on interface's methods during type parsing and add one more method to JsonMapper interface - arrayOf that must return proper Type for Array.
What do you think?

File: core/src/main/java/com/epam/coroutinecache/core/DiskCache.kt (hunk 1 above)
Comment (LikeAnyan) [reply core/src/main/java/com/epam/coroutinecache/core/DiskCache.kt]: oh wait you mean dividing by 1024, ye?

**Comment by LikeAnyan** [comment]
Pushed two commits over comments above, please write opinion about JsonMapper interface (I think that newParameterizedType could be removed from interface, but maybe there are cases when user needs it)

**Review by vvsevolodovich — APPROVED** [review]

**Pull request merged**